File: services/backend-go/internal/m29/pipeline/relations.go
package pipeline

import (
	"fmt"
	"math"

	"github.com/luqing-studio/image-figma/services/backend-go/internal/m29/contract"
)

func buildRelations(primitives []contract.Primitive) []contract.PhysicalRelation {
	var out []contract.PhysicalRelation
	nextID := 1
	for i := range primitives {
		for j := range primitives {
			if i == j {
				continue
			}
			a := primitives[i]
			b := primitives[j]
			if contains(a.BBox, b.BBox, 2) && area(a.BBox) > area(b.BBox) {
				out = append(out, contract.PhysicalRelation{
					ID:     fmt.Sprintf("rel_%04d", nextID),
					Kind:   "contains_bbox",
					FromID: a.ID,
					ToID:   b.ID,
					Ratio:  intersectionRatio(a.BBox, b.BBox),
				})
				nextID++
				continue
			}
			if a.PrimitiveType != "text_region" && b.PrimitiveType == "text_region" {
				d := bboxDistance(a.BBox, b.BBox)
				if d <= 12 {
					out = append(out, contract.PhysicalRelation{
						ID:       fmt.Sprintf("rel_%04d", nextID),
						Kind:     "near_text",
						FromID:   a.ID,
						ToID:     b.ID,
						Distance: d,
					})
					nextID++
				}
			}
		}
	}
	return out
}

func contains(parent, child contract.BBox, tolerance int) bool {
	return parent.X-tolerance <= child.X &&
		parent.Y-tolerance <= child.Y &&
		parent.X+parent.Width+tolerance >= child.X+child.Width &&
		parent.Y+parent.Height+tolerance >= child.Y+child.Height
}

func area(b contract.BBox) int {
	return max(0, b.Width) * max(0, b.Height)
}

func intersectionRatio(a, b contract.BBox) float64 {
	x1 := max(a.X, b.X)
	y1 := max(a.Y, b.Y)
	x2 := min(a.X+a.Width, b.X+b.Width)
	y2 := min(a.Y+a.Height, b.Y+b.Height)
	inter := area(contract.BBox{X: x1, Y: y1, Width: x2 - x1, Height: y2 - y1})
	if area(b) == 0 {
		return 0
	}
	return math.Round(float64(inter)/float64(area(b))*10000) / 10000
}

func bboxDistance(a, b contract.BBox) float64 {
	ax2, ay2 := a.X+a.Width, a.Y+a.Height
	bx2, by2 := b.X+b.Width, b.Y+b.Height
	dx := max(max(b.X-ax2, a.X-bx2), 0)
	dy := max(max(b.Y-ay2, a.Y-by2), 0)
	return math.Round(math.Sqrt(float64(dx*dx+dy*dy))*100) / 100
}

File: apps/slice-studio/server/m29-physical-evidence/relations.ts
import type { BBox } from "../../shared/types";
import type { M29PhysicalRelation, M29Primitive } from "./types";

export function buildRelations(primitives: M29Primitive[]): M29PhysicalRelation[] {
  const relations: M29PhysicalRelation[] = [];
  let nextId = 1;
  for (let i = 0; i < primitives.length; i += 1) {
    for (let j = 0; j < primitives.length; j += 1) {
      if (i === j) continue;
      const a = primitives[i];
      const b = primitives[j];
      if (contains(a.bbox, b.bbox, 2) && area(a.bbox) > area(b.bbox)) {
        relations.push({
          id: `rel_${String(nextId).padStart(4, "0")}`,
          kind: "contains_bbox",
          fromId: a.id,
          toId: b.id,
          ratio: intersectionRatio(a.bbox, b.bbox)
        });
        nextId += 1;
        continue;
      }
      if (a.primitiveType !== "text_region" && b.primitiveType === "text_region") {
        const distance = bboxDistance(a.bbox, b.bbox);
        if (distance <= 12) {
          relations.push({
            id: `rel_${String(nextId).padStart(4, "0")}`,
            kind: "near_text",
            fromId: a.id,
            toId: b.id,
            distance
          });
          nextId += 1;
        }
      }
    }
  }
  return relations;
}

function contains(parent: BBox, child: BBox, tolerance: number): boolean {
  return parent.x - tolerance <= child.x &&
    parent.y - tolerance <= child.y &&
    parent.x + parent.width + tolerance >= child.x + child.width &&
    parent.y + parent.height + tolerance >= child.y + child.height;
}

function area(bbox: BBox): number {
  return Math.max(0, bbox.width) * Math.max(0, bbox.height);
}

function intersectionRatio(a: BBox, b: BBox): number {
  const x1 = Math.max(a.x, b.x);
  const y1 = Math.max(a.y, b.y);
  const x2 = Math.min(a.x + a.width, b.x + b.width);
  const y2 = Math.min(a.y + a.height, b.y + b.height);
  const intersection = area({ x: x1, y: y1, width: x2 - x1, height: y2 - y1 });
  if (area(b) === 0) return 0;
  return Math.round((intersection / area(b)) * 10000) / 10000;
}

function bboxDistance(a: BBox, b: BBox): number {
  const ax2 = a.x + a.width;
  const ay2 = a.y + a.height;
  const bx2 = b.x + b.width;
  const by2 = b.y + b.height;
  const dx = Math.max(Math.max(b.x - ax2, a.x - bx2), 0);
  const dy = Math.max(Math.max(b.y - ay2, a.y - by2), 0);
  return Math.round(Math.sqrt(dx * dx + dy * dy) * 100) / 100;
}
